core/src/main/scala/replpp/util/ClasspathHelper.scala

```diff
@@ -0,0 +1,113 @@
+
+  /**
+   * Concatenates the classpath from multiple sources, each are required for different scenarios:
+   * - `java.class.path` system property
+   * - dependency artifacts as passed via (command-line) configuration
+   * - jars from current class loader (recursively)
+   *
+   * To have reproducable results, we order the classpath entries. This may interfere with the user's deliberate choice
+   * of order, but since we need to concatenate the classpath from different sources, the user can't really depend on
+   * the order anyway.
+   */
+  def create(config: Config, quiet: Boolean = false): String = {
+    /** Important: we absolutely have to make sure this starts and ends with a `pathSeparator`.
+     * Otherwise, the last entry is lost somewhere down the line (I didn't find the exact location where things go
+     * wrong, but it looked like somewhere in dotty 3.3.0). */
+    createAsSeq(config, quiet).mkString(pathSeparator, pathSeparator, pathSeparator)
+  }
+
+  protected[util] def createAsSeq(config: Config, quiet: Boolean = false): Seq[Path] = {
+    val entries = Seq.newBuilder[Path]
+    System.getProperty("java.class.path").split(pathSeparator).map(Paths.get(_)).foreach(entries.addOne)
+
+    val fromDependencies = dependencyArtifacts(config)
+    fromDependencies.foreach(entries.addOne)
+    if (fromDependencies.nonEmpty && !quiet) {
+      println(s"resolved dependencies - adding ${fromDependencies.size} artifact(s) to classpath - to list them, enable verbose mode")
+      if (verboseEnabled(config)) fromDependencies.foreach(println)
+    }
+
+    jarsFromClassLoaderRecursively(classOf[replpp.ReplDriver].getClassLoader)
+      .foreach(url => entries.addOne(Paths.get(url.toURI)))
+
+    onlyHighestVersionForEachJar(entries.result().map(versionInfoFromJar)).sorted
+  }
+
+  /**
+   * Goal: for each organisation/name tuple: filter out anything but the highest version. This is to ensure we don't have jars with different versions on the classpath.
+   * Problem: since we don't only get the jars via maven coordinates, but also inherit from existing classpath entries, all we have is a list of jars.
+   * The best solution is probably to use separate classloaders for the REPL itself and the code that's executed within. Similar to what ammonite does..
+   * In the interim, we use this hack as a semi-best-effort basis: group by the information in their manifests. */
+  private[util] def onlyHighestVersionForEachJar(dependencyInfos: Seq[JarWithManifestInfo]): Seq[Path] = {
+    dependencyInfos
+      .groupMap(dep => (dep.organisation, dep.name))(dep => (dep.jar, Version(dep.version.getOrElse("0"))))
+      .map { case ((organisation, name), jars) =>
+        if (jars.size == 1) {
+          jars.head._1
+        } else {
+          // choose the jar with the highest version
+          val result = jars.maxBy { case (_, version) => version } match {
+            case (jar, _) => jar
+          }
+          if (jars.size > 1) {
```

Comment:
this `if` seems redundant since it's in the `else` branch for `jars.size == 1` (and count 0 should be impossible after grouping). Personally, I'd probably drop that outer `if`...

Reply:
whoops, that was my intention actually, got lost somewhere on the way. thanks for catching